File: runtime/dart/typed_data/float64x2.ts
import { IFloat32x4 } from "./float32x4";
declare const dart: {
    typed_data: {
        float64x2: (
            this: void,
            float64x2: IFloat64x2,
            x: number,
            y: number
        ) => IFloat64x2;
        float64x2Splat: (v: number) => IFloat64x2;
        float64x2Zero: () => IFloat64x2;
        float64x2FromFloat32x4: (v: IFloat32x4) => IFloat64x2;
    };
};
export interface IFloat64x2 {
    scale: (s: number) => IFloat64x2;
    abs: () => IFloat64x2;
    clamp: (lowerLimit: IFloat64x2, upperLimit: IFloat64x2) => IFloat64x2;
    getX: () => number;
    getY: () => number;
    getSignMask: () => number;
    withX: (x: number) => IFloat64x2;
    withY: (y: number) => IFloat64x2;
    min: (other: IFloat64x2) => IFloat64x2;
    max: (other: IFloat64x2) => IFloat64x2;
    sqrt: () => IFloat64x2;
}
export class Float64x2 {
    public constructor(x: number, y: number) {
        dart.typed_data.float64x2(this, x, y);
    }
    public static splat(v: number): IFloat64x2 {
        return dart.typed_data.float64x2Splat(v);
    }
    public static zero(): IFloat64x2 {
        return dart.typed_data.float64x2Zero();
    }
    public static fromFloat32x4(v: IFloat32x4): IFloat64x2 {
        return dart.typed_data.float64x2FromFloat32x4(v);
    }
    private readonly _dart_scale: (s: number) => IFloat64x2 = undefined as any;
    private readonly _dart_abs: () => IFloat64x2 = undefined as any;
    private readonly _dart_clamp: (
        lowerLimit: IFloat64x2,
        upperLimit: IFloat64x2
    ) => IFloat64x2 = undefined as any;
    private readonly _dart_getX: () => number = undefined as any;
    private readonly _dart_getY: () => number = undefined as any;
    private readonly _dart_getSignMask: () => number = undefined as any;
    private readonly _dart_withX: (x: number) => IFloat64x2 = undefined as any;
    private readonly _dart_withY: (y: number) => IFloat64x2 = undefined as any;
    private readonly _dart_min: (
        other: IFloat64x2
    ) => IFloat64x2 = undefined as any;
    private readonly _dart_max: (
        other: IFloat64x2
    ) => IFloat64x2 = undefined as any;
    private readonly _dart_sqrt: () => IFloat64x2 = undefined as any;
    public scale(s: number): IFloat64x2 {
        return this._dart_scale(s);
    }
    public abs(): IFloat64x2 {
        return this._dart_abs();
    }
    public clamp(lowerLimit: IFloat64x2, upperLimit: IFloat64x2): IFloat64x2 {
        return this._dart_clamp(lowerLimit, upperLimit);
    }
    public getX(): number {
        return this._dart_getX();
    }
    public getY(): number {
        return this._dart_getY();
    }
    public getSignMask(): number {
        return this._dart_getSignMask();
    }
    public withX(x: number): IFloat64x2 {
        return this._dart_withX(x);
    }
    public withY(y: number): IFloat64x2 {
        return this._dart_withY(y);
    }
    public min(other: IFloat64x2): IFloat64x2 {
        return this._dart_min(other);
    }
    public max(other: IFloat64x2): IFloat64x2 {
        return this._dart_max(other);
    }
    public sqrt(): IFloat64x2 {
        return this._dart_sqrt();
    }
}

File: lib/cfr/builtins/libs/dart/typed_data/float64x2.dart
import 'dart:core';
import 'dart:typed_data';

import 'package:hydro_sdk/cfr/runtimeSupport.dart';

class VMManagedFloat64x2 extends VMManagedBox<Float64x2> {
  VMManagedFloat64x2(
      {required this.table, required this.vmObject, required this.hydroState})
      : super(
          table: table,
          vmObject: vmObject,
          hydroState: hydroState,
        ) {
    table['scale'] = makeLuaDartFunc(func: (List<dynamic> luaCallerArguments) {
      return [
        maybeBoxObject<Float64x2>(
            object: vmObject.scale(luaCallerArguments[1]?.toDouble()),
            hydroState: hydroState,
            table: HydroTable()),
      ];
    });
    table['abs'] = makeLuaDartFunc(func: (List<dynamic> luaCallerArguments) {
      return [
        maybeBoxObject<Float64x2>(
            object: vmObject.abs(),
            hydroState: hydroState,
            table: HydroTable()),
      ];
    });
    table['clamp'] = makeLuaDartFunc(func: (List<dynamic> luaCallerArguments) {
      return [
        maybeBoxObject<Float64x2>(
            object: vmObject.clamp(
                maybeUnBoxAndBuildArgument<Float64x2, dynamic>(
                    luaCallerArguments[1],
                    parentState: hydroState),
                maybeUnBoxAndBuildArgument<Float64x2, dynamic>(
                    luaCallerArguments[2],
                    parentState: hydroState)),
            hydroState: hydroState,
            table: HydroTable()),
      ];
    });
    table['getX'] = makeLuaDartFunc(func: (List<dynamic> luaCallerArguments) {
      return [
        vmObject.x,
      ];
    });
    table['getY'] = makeLuaDartFunc(func: (List<dynamic> luaCallerArguments) {
      return [
        vmObject.y,
      ];
    });
    table['getSignMask'] =
        makeLuaDartFunc(func: (List<dynamic> luaCallerArguments) {
      return [
        vmObject.signMask,
      ];
    });
    table['withX'] = makeLuaDartFunc(func: (List<dynamic> luaCallerArguments) {
      return [
        maybeBoxObject<Float64x2>(
            object: vmObject.withX(luaCallerArguments[1]?.toDouble()),
            hydroState: hydroState,
            table: HydroTable()),
      ];
    });
    table['withY'] = makeLuaDartFunc(func: (List<dynamic> luaCallerArguments) {
      return [
        maybeBoxObject<Float64x2>(
            object: vmObject.withY(luaCallerArguments[1]?.toDouble()),
            hydroState: hydroState,
            table: HydroTable()),
      ];
    });
    table['min'] = makeLuaDartFunc(func: (List<dynamic> luaCallerArguments) {
      return [
        maybeBoxObject<Float64x2>(
            object: vmObject.min(maybeUnBoxAndBuildArgument<Float64x2, dynamic>(
                luaCallerArguments[1],
                parentState: hydroState)),
            hydroState: hydroState,
            table: HydroTable()),
      ];
    });
    table['max'] = makeLuaDartFunc(func: (List<dynamic> luaCallerArguments) {
      return [
        maybeBoxObject<Float64x2>(
            object: vmObject.max(maybeUnBoxAndBuildArgument<Float64x2, dynamic>(
                luaCallerArguments[1],
                parentState: hydroState)),
            hydroState: hydroState,
            table: HydroTable()),
      ];
    });
    table['sqrt'] = makeLuaDartFunc(func: (List<dynamic> luaCallerArguments) {
      return [
        maybeBoxObject<Float64x2>(
            object: vmObject.sqrt(),
            hydroState: hydroState,
            table: HydroTable()),
      ];
    });
  }

  final HydroTable table;

  final HydroState hydroState;

  final Float64x2 vmObject;
}

void loadFloat64x2(
    {required HydroState hydroState, required HydroTable table}) {
  table['float64x2'] =
      makeLuaDartFunc(func: (List<dynamic> luaCallerArguments) {
    return [
      maybeBoxObject<Float64x2>(
          object: Float64x2(luaCallerArguments[1]?.toDouble(),
              luaCallerArguments[2]?.toDouble()),
          hydroState: hydroState,
          table: luaCallerArguments[0])
    ];
  });
  table['float64x2Splat'] =
      makeLuaDartFunc(func: (List<dynamic> luaCallerArguments) {
    return [
      maybeBoxObject<Float64x2>(
          object: Float64x2.splat(luaCallerArguments[1]?.toDouble()),
          hydroState: hydroState,
          table: HydroTable()),
    ];
  });
  table['float64x2Zero'] =
      makeLuaDartFunc(func: (List<dynamic> luaCallerArguments) {
    return [
      maybeBoxObject<Float64x2>(
          object: Float64x2.zero(),
          hydroState: hydroState,
          table: HydroTable()),
    ];
  });
  table['float64x2FromFloat32x4'] =
      makeLuaDartFunc(func: (List<dynamic> luaCallerArguments) {
    return [
      maybeBoxObject<Float64x2>(
          object: Float64x2.fromFloat32x4(
              maybeUnBoxAndBuildArgument<Float32x4, dynamic>(
                  luaCallerArguments[1],
                  parentState: hydroState)),
          hydroState: hydroState,
          table: HydroTable()),
    ];
  });
  registerBoxer<Float64x2>(boxer: (
      {required Float64x2 vmObject,
      required HydroState hydroState,
      required HydroTable table}) {
    return VMManagedFloat64x2(
        vmObject: vmObject, hydroState: hydroState, table: table);
  });
}
